datasimulator/node.py

```diff
@@ -0,0 +1,342 @@
+    generate_string_data,
+    generate_simple_primitive_data,
+)
+from utils import is_mixed_type, random_choice, get_recursive_keys
+
+
+# Ingnore system properties
+EXCLUDED_FIELDS = [
+    "type",
+    "error_type",
+    "state",
+    "id",
+    "file_state",
+    "state",
+    "state_comment",
+    "project_id",
+    "submitter_id",
+    "object_id",
+]
+
+
+logger = get_logger("DataSimulator")
+
+
+class Node(object):
+    """
+    Class representation for node
+    """
+
+    def __init__(self, node_name, node_schema, project):
+        """
+        """
+        self.name = node_name
+        self.project = project
+        self.required = node_schema.get("required", [])
+        self.sys_properties = node_schema.get("systemProperties", [])
+        try:
+            self.category = node_schema["category"]
+            self.properties = node_schema["properties"]
+            self.links = node_schema["links"]
+        except KeyError as e:
+            raise UserError(
+                "Error: NODE {} does not have key `{}`".format(node_name, e.message)
+            )
+        self.required_links = []
+        self.simulated_dataset = []
+
+    def __str__(self):
+        return self.name
+
+    @staticmethod
+    def _is_link_property(prop_schema):
+        keys = get_recursive_keys(prop_schema)
+        if "id" in keys and "pattern" in keys and "term" in keys:
+            return True
+        return False
+
+    @staticmethod
+    def _is_datetime_property(prop_schema):
+        if str(prop_schema).find("date and time of day") > 0:
+            return True
+        return False
+
+    @staticmethod
+    def _simulate_data_from_simple_schema(prop, simple_schema):
+        if simple_schema["data_type"] == "md5sum":
+            return generate_hash()
+        elif simple_schema["data_type"] == "enum":
+            return random_choice(simple_schema["values"])
+        elif simple_schema["data_type"] == "datetime":
+            return generate_datetime()
+        else:
+            return generate_simple_primitive_data(
+                data_type=simple_schema["data_type"],
+                pattern=simple_schema.get("pattern"),
+            )
+
+    def node_validation(self, required_only=False):
+        """
+        Validate the node schema and check if the node is submitable or not.
+        The method supports for generating data even when the node does not pass validation.
+        Some non-required properties, which can be missed, may fail the validation.
+
+        Args:
+            required_only(bool)
+
+        Outputs:
+            pass_validation(bool): The node does not pass validation
+            is_submitable(bool): The node can be submitted even it does not pass validation
+        """
+        # if node schema is well-defined
+        pass_validation = True
+        is_submittable = True
+
+        # Go through all the properties to collect all possible errors may happend
+        for prop in self.required:
```

Comment:
```suggestion
        if any(prop not in self.properties for prop in self.required):
            # check which one failed
```

Reply:
or:
```suggestion
        if not set(self.required) <= set(self.properties):
```

Reply:
The fist option is better since we don't have to convert to set that results in more computation

Reply:
The logic of both the suggestion seems wrong? :).

Reply:
```
In [1]: xs = [1, 3, 5, 6]

In [2]: ys = [1, 4, 7]

In [3]: required = [1, 3, 6]

In [4]: if any(r not in xs for r in required):
   ...:     print("xs fail")
   ...:     

In [5]: if any(r not in ys for r in required):
   ...:     print("ys fail")
   ...:     
ys fail

In [6]: set(required) <= set(xs)
Out[6]: True

In [7]: set(required) <= set(ys)
Out[7]: False
```

Reply:
If you want all the elements in `self.required` that are not in `self.properties`, do:
```python
set(self.required) - set(self.properties)
```
which will be a set of elements in required that are not in properties.

Reply:
Yeah. We can do `set(self.required) - set(self.properties)` to get all the required that is not in properties but we need to convert list to set. It requires more computation right? Now I see what do you meant with the comment `# check which one failed`